File: ast/ast_tester/testcmpframe.c
/*
 *  Test the CmpFrame class (NormPoints).
 *  Converted from the Fortran test testcmpframe.f.
 *  Direct conversion; no material differences from the Fortran original.
 */
#include "ast.h"
#include <stdio.h>

static void stopit( int *status, const char *text ) {
   if( *status != 0 ) return;
   *status = 1;
   printf( "%s\n", text );
}

static void fill_input( double in[ static 12 ], const double ra[ static 4 ],
                        const double px[ static 4 ], const double dec[ static 4 ] ) {
   int i;
   for( i = 0; i < 4; i++ ) {
      in[ i ] = ra[ i ];
      in[ i + 4 ] = px[ i ];
      in[ i + 8 ] = dec[ i ];
   }
}

static int check_normed_output( const double out[ static 12 ],
                                const double ra[ static 4 ],
                                const double px[ static 4 ],
                                const double dec[ static 4 ] ) {
   int i;
   if( out[ 0 ] != ra[ 0 ] - 2*AST__DPI ) return 1;
   if( out[ 1 ] != ra[ 1 ] - 2*AST__DPI ) return 2;
   if( out[ 2 ] != ra[ 2 ] ) return 3;
   if( out[ 3 ] != ra[ 3 ] ) return 4;
   for( i = 0; i < 4; i++ ) {
      if( out[ i + 4 ] != px[ i ] ) return 5;
      if( out[ i + 8 ] != dec[ i ] ) return 6;
   }
   return 0;
}

static int check_passthrough_output( const double out[ static 12 ],
                                     const double ra[ static 4 ],
                                     const double px[ static 4 ],
                                     const double dec[ static 4 ] ) {
   int i;
   for( i = 0; i < 4; i++ ) {
      if( out[ i ] != ra[ i ] ) return 7;
      if( out[ i + 4 ] != px[ i ] ) return 8;
      if( out[ i + 8 ] != dec[ i ] ) return 9;
   }
   return 0;
}

int main( void ) {
   int status_value = 0;
   int *status = &status_value;
   int perm[3];
   AstSkyFrame *sfrm;
   AstFrame *pfrm;
   AstCmpFrame *cfrm;

   double ra[]  = { 6.1, 6.1, 0.04, 0.04 };
   double dec[] = { 0.2, -0.2, -0.2, 0.2 };
   double px[]  = { -100.0, -10.0, 10.0, 100.0 };
   double in[4*3], out[4*3];

   astWatch( status );
   astBegin;

   sfrm = astSkyFrame( " " );
   perm[0] = 2;
   perm[1] = 1;
   astPermAxes( sfrm, perm );

   pfrm = astFrame( 1, "Domain=FPLANE" );

   cfrm = astCmpFrame( pfrm, sfrm, " " );
   perm[0] = 3;
   perm[1] = 1;
   perm[2] = 2;
   astPermAxes( cfrm, perm );

   /* Fill input array: columns are RA, PX, Dec (after permutation).
      Fortran layout in(4,3) with column-major ordering becomes
      a flat array in[npoint*ncoord] with npoint=4, ncoord=3.
      in[i + npoint*coord] = value for point i, coordinate coord. */
   fill_input( in, ra, px, dec );

   /* Normalise with `contig=1` (coordinates are contiguous per axis). */
   astNormPoints( cfrm, 4, 3, 4, in, 1, 3, 4, out );

   switch( check_normed_output( out, ra, px, dec ) ) {
   case 1: stopit( status, "Error 1" ); break;
   case 2: stopit( status, "Error 2" ); break;
   case 3: stopit( status, "Error 3" ); break;
   case 4: stopit( status, "Error 4" ); break;
   case 5: stopit( status, "Error 5" ); break;
   case 6: stopit( status, "Error 6" ); break;
   }

   /* Normalise with `contig=0` (do not normalise). */
   astNormPoints( cfrm, 4, 3, 4, in, 0, 3, 4, out );

   switch( check_passthrough_output( out, ra, px, dec ) ) {
   case 7: stopit( status, "Error 7" ); break;
   case 8: stopit( status, "Error 8" ); break;
   case 9: stopit( status, "Error 9" ); break;
   }

   astEnd;
   astFlushMemory( 1 );

   if( *status == 0 ) {
      printf( " All CmpFrame tests passed\n" );
   } else {
      printf( "CmpFrame tests failed\n" );
   }
   return *status;
}

File: ast/ast_tester/testcmpframe.f
      program testcmpframe
      implicit none

      include 'AST_PAR'
      include 'AST_ERR'
      include 'SAE_PAR'

      integer status, pfrm, sfrm, cfrm, perm(3), nc, i
      double precision in( 4, 3 ), ra(4), dec(4), px(4)
      double precision out( 4, 3 )

      data ra / 6.1D0, 6.1D0, 0.04D0, 0.04D0 /,
     :     dec / 0.2D0, -0.2D0, -0.2D0, 0.2D0 /,
     :     px / -100.0D0, -10.0D0, 10.0D0, 100.0D0 /

c      call ast_watchmemory(100)

      status = sai__ok
      call err_mark( status )
      call ast_begin( status )


      sfrm = ast_skyframe( ' ', status )
      perm( 1 ) = 2
      perm( 2 ) = 1
      call ast_permaxes( sfrm, perm, status )

      pfrm = ast_frame( 1, "Domain=FPLANE", status )


      cfrm = ast_cmpframe( pfrm, sfrm, ' ', status )
      perm( 1 ) = 3
      perm( 2 ) = 1
      perm( 3 ) = 2
      call ast_permaxes( cfrm, perm, status )

      do i = 1, 4
         in( i, 1 ) = ra( i )
         in( i, 2 ) = px( i )
         in( i, 3 ) = dec( i )
      end do

      call ast_normpoints( cfrm, 4, 3, 4, in, 1, 3, 4, out, status )

      if( out( 1, 1 ) .ne. ra( 1 ) - 2*AST__DPI ) then
         call stopit( status, 'Error 1' )
      else if( out( 2, 1 ) .ne. ra( 2 ) - 2*AST__DPI ) then
         call stopit( status, 'Error 2' )
      else if( out( 3, 1 ) .ne. ra( 3 ) ) then
         call stopit( status, 'Error 3' )
      else if( out( 4, 1 ) .ne. ra( 4 ) ) then
         call stopit( status, 'Error 4' )
      else
         do i = 1, 4
            if( out( i, 2 ) .ne. px( i ) ) then
               call stopit( status, 'Error 5' )
            else if( out( i, 3 ) .ne. dec( i ) ) then
               call stopit( status, 'Error 6' )
            end if
         end do
      end if

      call ast_normpoints( cfrm, 4, 3, 4, in, 0, 3, 4, out, status )

      do i = 1, 4
         if( out( i, 1 ) .ne. ra( i ) ) then
            call stopit( status, 'Error 7' )
         else if( out( i, 2 ) .ne. px( i ) ) then
            call stopit( status, 'Error 8' )
         else if( out( i, 3 ) .ne. dec( i ) ) then
            call stopit( status, 'Error 9' )
         end if
      end do

      call ast_end( status )
      call err_rlse( status )

      call ast_activememory( 'testcmpframe' )
      call ast_flushmemory( 1 );

      if( status .eq. sai__ok ) then
         write(*,*) 'All CmpFrame tests passed'
      else
         write(*,*) 'CmpFrame tests failed'
      end if

      end



      subroutine stopit( status, text )
      implicit none
      include 'SAE_PAR'
      integer status
      character text*(*)

      if( status .ne. sai__ok ) return
      status = sai__error
      write(*,*) text

      end

